CLI: DRY up how the CLI shows deployment and release URLs

vitorfhc opened the pull request:
Closes #4209

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/waypoint?pullRequest=4399)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/waypoint?pullRequest=4399)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)


---
**Vitor Falcao** seems not to be a GitHub user.
You need a GitHub account to be able to sign the CLA. If you already have a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/waypoint?pullRequest=4399).</sub>

vitorfhc commented on internal/cli/helpers.go:
```diff
@@ -0,0 +1,36 @@
+		return urlString, nil
+	}
+
+	// Avoids adding a scheme if it already has a different one
+	if parsedUrl.Scheme != "" && parsedUrl.Scheme != string(scheme) {
+		err := fmt.Errorf("The URL \"%s\" already has a different scheme: \"%s\"", urlString, scheme)
+		return "", err
+	}
```
Not sure if it makes sense

briancain commented on internal/cli/helpers.go:
```diff
@@ -0,0 +1,36 @@
+package cli
```
We generally try to stay away from file names like `helpers` or `utils`. I would simply call this file for what it is: `internal/cli/url_scheme.go`

briancain commented on internal/cli/helpers.go:
```diff
@@ -0,0 +1,36 @@
+
+	// Avoids adding a scheme if it already has a different one
+	if parsedUrl.Scheme != "" && parsedUrl.Scheme != string(scheme) {
+		err := fmt.Errorf("The URL \"%s\" already has a different scheme: \"%s\"", urlString, scheme)
```
```suggestion
		err := fmt.Errorf("The URL %q already has a different scheme: %q", urlString, scheme)
```

`%q` is the same thing as `\"%s\"`

briancain commented on internal/cli/helpers.go:
```diff
@@ -0,0 +1,36 @@
+	}
+
+	parsedUrl.Scheme = string(scheme)
+	urlWithScheme := fmt.Sprintf("%s", parsedUrl)
```
This looks to be a linter failure:

```
internal/cli/helpers.go:34:19: S1025: should use String() instead of fmt.Sprintf (gosimple)
	urlWithScheme := fmt.Sprintf("%s", parsedUrl)
```

vitorfhc commented:
@briancain I deleted the `closes` on the description since I won't have enough time to DRY up everything stated in issue #4209. I changed the PR to Ready to Review.

cicoyle commented on internal/cli/url_scheme.go:
```diff
@@ -0,0 +1,36 @@
+
+	// Avoids adding a scheme if it already has a different one
+	if parsedUrl.Scheme != "" && parsedUrl.Scheme != string(scheme) {
+		err := fmt.Errorf("The URL \"%s\" already has a different scheme: \"%s\"", urlString, scheme)
```
```suggestion
		err := fmt.Errorf("The URL %q already has a different scheme: %q", urlString, scheme)
```

cicoyle commented on internal/cli/url_scheme.go:
```diff
@@ -0,0 +1,36 @@
+		return "", err
+	}
+
+	if parsedUrl.Scheme == string(scheme) || urlString == "" {
```
We might not need the or case because we won't want to return the empty urlString

cicoyle replied:
It might be worth explicitly checking this separately in its own case. And if we do want this check to be an ok case, we might want to move it up before the url parsing. In any case, it should match the current behavior.

cicoyle reviewed (commented):
This is a great idea! Thanks for working on it. I believe there may be a missing file change for the `waypoint release` case.

cicoyle commented on internal/cli/url_scheme.go:
```diff
@@ -0,0 +1,36 @@
+	httpsScheme UrlScheme = "https"
+	httpScheme  UrlScheme = "http"
+)
+
```
Mind adding some documentation on how to use this function and the expected format is for the url string? I think that additional insight would be valuable to have - even a snip it of such examples as a comment.

cicoyle commented on internal/cli/url_scheme.go:
```diff
@@ -0,0 +1,36 @@
+	parsedUrl, err := url.Parse(urlString)
+	// The url.Parse function may return an error when there's no scheme due to parser ambiguities,
+	// so we need to verify if the error is because the url lacks a scheme
+	if err != nil && parsedUrl.Scheme != "" {
```
Mind pulling out the err check and delete the rest?
```suggestion
	if err != nil {
```

xiaolin-ninja commented:
Hi @vitorfhc, are you still working on this PR?
